From 0d835cda433b42d2e8201e77e5cefd79bff34d32 Mon Sep 17 00:00:00 2001
From: SuicidalKids <kfian294ma4@gmail.com>
Date: Thu, 29 Oct 2020 21:09:34 +0000
Subject: [PATCH] Disable Mob Sounds


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index be66ad4c3..04022cfd0 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -107,4 +107,11 @@ public class IonWorldConfig {
         requirePlayers = getBoolean("spawners.nearby.require-players", true);
     }
 
+    public boolean footstepSounds;
+    public boolean silenceSounds;
+    private void Sounds() {
+        footstepSounds = getBoolean("footstep-sounds", true);
+        silenceSounds = getBoolean("silence-sounds", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 9d95e8570..14fdd54b1 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -789,7 +789,11 @@ public abstract class Entity implements ICommandListener {
                         this.makeSound(this.P(), f, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
                     }
 
+                    // IonSpigot start - Disable footstep sounds
+                    if (world.ionConfig.footstepSounds) {
                     this.a(blockposition, block);
+                    }
+                    // Ionspigot end
                     block.a(this.world, blockposition, this); // CraftBukkit moved from above
                 }
             }
@@ -897,7 +901,7 @@ public abstract class Entity implements ICommandListener {
     }
 
     public boolean R() {
-        return this.datawatcher.getByte(4) == 1;
+        return !world.ionConfig.silenceSounds && this.datawatcher.getByte(4) == 1;
     }
 
     public void b(boolean flag) {
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 12257ad24..a6c167cae 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -155,7 +155,7 @@ public abstract class EntityLiving extends Entity {
                 // CraftBukkit start - visiblity api
                 if (this instanceof EntityPlayer) {
                     ((WorldServer) this.world).sendParticles((EntityPlayer) this, EnumParticle.BLOCK_DUST, false, this.locX, this.locY, this.locZ, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D, new int[] { Block.getCombinedId(iblockdata)});
-                } else {
+                } else if (!world.ionConfig.silenceSounds) { // IonSpigot
                     ((WorldServer) this.world).a(EnumParticle.BLOCK_DUST, this.locX, this.locY, this.locZ, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D, new int[] { Block.getCombinedId(iblockdata)});
                 }
                 // CraftBukkit end
@@ -960,7 +960,11 @@ public abstract class EntityLiving extends Entity {
             if (block.getMaterial() != Material.AIR) {
                 Block.StepSound block_stepsound = block.stepSound;
 
+                // IonSpigot start - Disable footstep sounds
+                if (world.ionConfig.footstepSounds) {
                 this.makeSound(block_stepsound.getStepSound(), block_stepsound.getVolume1() * 0.5F, block_stepsound.getVolume2() * 0.75F);
+                }
+                // IonSpigot end
             }
         }
 
-- 
2.29.2.windows.2

