From fd01a63c722cb7959d8bc3d66123fb00843033cc Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Fri, 27 Nov 2020 22:02:11 +0000
Subject: [PATCH] Merge Spawned Items


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index 67630ecc6..26b165719 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -128,4 +128,9 @@ public class IonWorldConfig {
         hopperInterval = getInterval("items.hoppers.delay", 0);
     }
 
+    public boolean mergeSpawnedItems;
+    private void MergeSpawnedItems() {
+        mergeSpawnedItems = getBoolean("items.merged-spawned-items", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 3fdb2709b..a3693edc8 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -54,6 +54,7 @@ public class Chunk {
     public AtomicInteger pendingLightUpdates = new AtomicInteger();
     public long lightUpdateTime;
     // PaperSpigot end
+    public EntityItem cachedEntityItem; // IonSpigot - Merge Spawned Items
 
     // PaperSpigot start - ChunkMap caching
     private PacketPlayOutMapChunk.ChunkMap chunkMap;
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 918c1bdef..90dfced16 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -167,7 +167,7 @@ public class EntityItem extends Entity implements HopperPusher {
 
     }
 
-    private boolean a(EntityItem entityitem) {
+    public boolean a(EntityItem entityitem) { // IonSpigot - private -> public
         if (entityitem == this) {
             return false;
         } else if (entityitem.isAlive() && this.isAlive()) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index f32c106d6..8a024afbf 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -998,7 +998,29 @@ public abstract class World implements IBlockAccess {
         return true;
     }
 
+    // IonSpigot start - Merge Spawned Items
+    public boolean addEntity(EntityItem entityItem) {
+        if (ionConfig.mergeSpawnedItems) {
+            int floorX = org.bukkit.util.NumberConversions.floor(entityItem.locX) >> 4;
+            int floorZ = org.bukkit.util.NumberConversions.floor(entityItem.locZ) >> 4;
+            Chunk chunk = getChunkIfLoaded(floorX, floorZ);
+
+            EntityItem cachedItem = chunk.cachedEntityItem;
+            if (cachedItem != null && cachedItem.isAlive() &&
+                cachedItem.h(entityItem) < spigotConfig.itemMerge * spigotConfig.itemMerge &&
+                cachedItem.a(entityItem) && !entityItem.isAlive()) {
+                return false;
+            }
+
+            chunk.cachedEntityItem = entityItem;
+        }
+        return addEntity(entityItem, SpawnReason.DEFAULT);
+    }
     public boolean addEntity(Entity entity) {
+        if (entity instanceof EntityItem) {
+            return addEntity((EntityItem) entity);
+        }
+        // IonSpigot end
         // CraftBukkit start - Used for entities other than creatures
         return addEntity(entity, SpawnReason.DEFAULT);
     }
-- 
2.29.2.windows.2

