From f309d2954b9eca53cd4e4cbebee3c959962ca12c Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Sun, 29 Nov 2020 20:24:25 +0000
Subject: [PATCH] Nerf Spawned Mob Logic


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index 73095aeab..56ef01962 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -121,4 +121,11 @@ public class IonWorldConfig {
         mergeSpawnedItems = getBoolean("items.merged-spawned-items", false);
     }
 
+    public boolean nerfSpawnedLogic;
+    public boolean nerfNaturalSpawns;
+    private void NerfSpawnedLogic() {
+        nerfSpawnedLogic = getBoolean("spawners.nerf-spawned-logic", false);
+        nerfNaturalSpawns = getBoolean("nerf-natural-spawns", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 14fdd54b1..6d4ec0500 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -146,6 +146,7 @@ public abstract class Entity implements ICommandListener {
     public boolean fromMobSpawner;
     public void inactiveTick() { }
     // Spigot end
+    public boolean naturalSpawn; // IonSpigot
 
     public int getId() {
         return this.id;
@@ -1323,6 +1324,8 @@ public abstract class Entity implements ICommandListener {
                 }
             }
 
+            nbttagcompound.setBoolean("FromMobSpawner", this.fromMobSpawner); // IonSpigot
+            nbttagcompound.setBoolean("NaturalSpawn", this.naturalSpawn); // IonSpigot
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.a(throwable, "Saving entity NBT");
             CrashReportSystemDetails crashreportsystemdetails = crashreport.a("Entity being saved");
@@ -1389,6 +1392,14 @@ public abstract class Entity implements ICommandListener {
             if (this.af()) {
                 this.setPosition(this.locX, this.locY, this.locZ);
             }
+            // IonSpigot start
+            if (nbttagcompound.hasKey("FromMobSpawner")) {
+                this.fromMobSpawner = nbttagcompound.getBoolean("FromMobSpawner");
+            }
+            if (nbttagcompound.hasKey("NaturalSpawn")) {
+                this.naturalSpawn = nbttagcompound.getBoolean("NaturalSpawn");
+            }
+            // IonSpigot end
 
             // CraftBukkit start
             if (this instanceof EntityLiving) {
diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index b82b98473..cd2307d21 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -459,7 +459,8 @@ public abstract class EntityInsentient extends EntityLiving {
         this.D();
         this.world.methodProfiler.b();
         // Spigot Start
-        if ( this.fromMobSpawner )
+        if ( this.fromMobSpawner && (world.spigotConfig.nerfSpawnerMobs || world.ionConfig.nerfSpawnedLogic) ||
+             this.naturalSpawn && world.ionConfig.nerfNaturalSpawns) // IonSpigot - Nerf Spawned Mob Logic
         {
             // PaperSpigot start - Allow nerfed mobs to jump
             if (goalFloat != null) {
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index a6c167cae..2bf8928e3 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -173,6 +173,21 @@ public abstract class EntityLiving extends Entity {
         this.ay = this.az;
         super.K();
         this.world.methodProfiler.a("livingEntityBaseTick");
+        // IonSpigot start - Nerf Spawned Mob Logic
+        if ((this.fromMobSpawner && world.ionConfig.nerfSpawnedLogic ||
+             this.naturalSpawn && world.ionConfig.nerfNaturalSpawns) &&
+            (!onGround || deathTicks != 0)) {
+            if (this.deathTicks != 0) {
+                this.aZ();
+            }
+
+            if (this.noDamageTicks > 0) {
+                --this.noDamageTicks;
+            }
+
+            return;
+        }
+        // IonSpigot end
         boolean flag = this instanceof EntityHuman;
 
         if (this.isAlive()) {
@@ -1470,6 +1485,13 @@ public abstract class EntityLiving extends Entity {
 
     public void t_() {
         super.t_();
+        // IonSpigot start - Nerf Spawned Mob Logic
+        if (this.fromMobSpawner && world.ionConfig.nerfSpawnedLogic ||
+            this.naturalSpawn && world.ionConfig.nerfNaturalSpawns) {
+            this.m();
+            return;
+        }
+        // IonSpigot end
         if (!this.world.isClientSide) {
             int i = this.bv();
 
@@ -1602,6 +1624,35 @@ public abstract class EntityLiving extends Entity {
     }
 
     public void m() {
+        // IonSpigot start - Nerf Spawned Mob Logic
+        if (this.fromMobSpawner && world.ionConfig.nerfSpawnedLogic ||
+            this.naturalSpawn && world.ionConfig.nerfNaturalSpawns) {
+            if (deathTicks == 0) {
+                this.doTick();
+
+                if (this.aY) {
+                    if (this.V()) {
+                        this.bG();
+                    } else if (this.ab()) {
+                        this.bH();
+                    } else if (this.onGround && this.bn == 0) {
+                        this.bF();
+                        this.bn = 10;
+                    }
+                } else {
+                    this.bn = 0;
+                }
+
+                this.move(this.motX, this.motY, this.motZ);
+                this.motY -= 0.08D;
+                this.motY *= 0.9800000190734863D;
+                this.motX *= 0.91F;
+                this.motZ *= 0.91F;
+            }
+            
+            return;
+        }
+        // IonSpigot end
         if (this.bn > 0) {
             --this.bn;
         }
diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index 2f8750a30..3bae8f6d1 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -167,7 +167,7 @@ public abstract class MobSpawnerAbstract {
                 if (!event.isCancelled()) {
                     entity.world.addEntity(entity, CreatureSpawnEvent.SpawnReason.SPAWNER); // CraftBukkit
                     // Spigot Start
-                    if ( entity.world.spigotConfig.nerfSpawnerMobs )
+                    if ( true ) // IonSpigot
                     {
                         entity.fromMobSpawner = true;
                     }
@@ -220,7 +220,7 @@ public abstract class MobSpawnerAbstract {
             if (!event.isCancelled()) {
                 entity.world.addEntity(entity, CreatureSpawnEvent.SpawnReason.SPAWNER); // CraftBukkit
                 // Spigot Start
-                if ( entity.world.spigotConfig.nerfSpawnerMobs )
+                if ( true ) // IonSpigot
                 {
                     entity.fromMobSpawner = true;
                 }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 8a024afbf..1525ea766 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1048,6 +1048,11 @@ public abstract class World implements IBlockAccess {
                     entity.dead = true;
                     return false;
                 }
+                // IonSpigot start
+                if (spawnReason != SpawnReason.SPAWNER && spawnReason != SpawnReason.SPAWNER_EGG) {
+                    entity.naturalSpawn = true;
+                }
+                // IonSpigot end
             }
 
             event = CraftEventFactory.callCreatureSpawnEvent((EntityLiving) entity, spawnReason);
-- 
2.29.2.windows.2

