From 8220f6a152f88ac895738c8efdec29d087b95907 Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Sun, 29 Nov 2020 20:26:27 +0000
Subject: [PATCH] Allow completely disabling pushing


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 3f4c6d473..e40f22c15 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1731,6 +1731,11 @@ public abstract class EntityLiving extends Entity {
     protected void doTick() {}
 
     protected void bL() {
+        // IonSpigot start - Allow completely disabling pushing
+        if (world.spigotConfig.maxCollisionsPerEntity == -1) {
+            return;
+        }
+        // IonSpigot end
         List list = this.world.a((Entity) this, this.getBoundingBox().grow(0.20000000298023224D, 0.0D, 0.20000000298023224D), Predicates.and(IEntitySelector.d, new Predicate() {
             public boolean a(Entity entity) {
                 return entity.ae();
-- 
2.29.2.windows.2

