From 1f3a552c182e4a720df61e36dbefe7ae51d79b50 Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Sun, 29 Nov 2020 20:42:34 +0000
Subject: [PATCH] Optional Movement Entity Collisions


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index 56ef01962..1cd8ef5f3 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -128,4 +128,9 @@ public class IonWorldConfig {
         nerfNaturalSpawns = getBoolean("nerf-natural-spawns", false);
     }
 
+    public boolean movementEntityCollisions;
+    private void MovementEntityCollisions() {
+        movementEntityCollisions = getBoolean("movement.entity-collisions", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1525ea766..719945a23 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1268,6 +1268,7 @@ public abstract class World implements IBlockAccess {
         }
         // Spigot end
 
+        if (!ionConfig.movementEntityCollisions) return arraylist; // IonSpigot - Optimise general movement...
         if (entity instanceof EntityItem) return arraylist; // PaperSpigot - Optimize item movement
         if (entity instanceof EntityArmorStand) return arraylist; // TacoSpigot - Optimize armor stand movement
         if (entity instanceof EntityTNTPrimed) return arraylist; // TacoSpigot - Optimize tnt entity movement
-- 
2.29.2.windows.2

