From fb03e1d87da684d3b4fbdd2b3f88e712cba2c37a Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Wed, 16 Dec 2020 18:26:12 +0000
Subject: [PATCH] Lag Compensated Potions


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index ac8491390..59980f163 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -130,4 +130,9 @@ public class IonWorldConfig {
         movementEntityCollisions = getBoolean("movement.entity-collisions", true);
     }
 
+    public boolean lagCompensatedPotions;
+    private void LagCompensatedPotions() {
+        lagCompensatedPotions = getBoolean("potions.lag-compensated", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index b029dbd55..f22fa3c83 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -53,6 +53,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public int ping;
     public boolean viewingCredits;
     public VisualSettings visualSettings = new VisualSettings(); // IonSpigot - Visual Settings API
+    public List<EntityPotion> potions = new ArrayList<>(); // IonSpigot - Lag Compensated Potions
 
     // CraftBukkit start
     public String displayName;
@@ -300,6 +301,21 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public void l() {
         try {
             super.t_();
+            // IonSpigot start - Lag Compensated Potions
+            for (int i = 0; i < this.potions.size(); ++i) {
+                EntityPotion entityPotion = this.potions.get(i);
+                entityPotion.tick();
+
+                // Check if size is > 9, this should cover some abuse
+                if (entityPotion.dead || this.potions.size() > 9) {
+                    if (!entityPotion.dead) {
+                        entityPotion.compensated = false;
+                    }
+
+                    this.potions.remove(i--);
+                }
+            }
+            // IonSpigot end
 
             for (int i = 0; i < this.inventory.getSize(); ++i) {
                 ItemStack itemstack = this.inventory.getItem(i);
diff --git a/src/main/java/net/minecraft/server/EntityPotion.java b/src/main/java/net/minecraft/server/EntityPotion.java
index fd174c346..2a1701c5a 100644
--- a/src/main/java/net/minecraft/server/EntityPotion.java
+++ b/src/main/java/net/minecraft/server/EntityPotion.java
@@ -12,6 +12,7 @@ import org.bukkit.entity.LivingEntity;
 
 public class EntityPotion extends EntityProjectile {
 
+    public boolean compensated = false; // IonSpigot - Lag Compensated Potions
     public ItemStack item;
 
     public EntityPotion(World world) {
@@ -25,6 +26,12 @@ public class EntityPotion extends EntityProjectile {
     public EntityPotion(World world, EntityLiving entityliving, ItemStack itemstack) {
         super(world, entityliving);
         this.item = itemstack;
+        // IonSpigot start - Lag Compensated Potions
+        if (entityliving instanceof EntityPlayer && world.ionConfig.lagCompensatedPotions) {
+            ((EntityPlayer) entityliving).potions.add(this);
+            compensated = true;
+        }
+        // IonSpigot end
     }
 
     public EntityPotion(World world, double d0, double d1, double d2, ItemStack itemstack) {
@@ -60,6 +67,17 @@ public class EntityPotion extends EntityProjectile {
         return this.item.getData();
     }
 
+    // IonSpigot start - Lag Compensated Potions
+    @Override
+    public void t_() {
+        if (!compensated) {
+            tick();
+        }
+    }
+    public void tick() {
+        super.t_();
+    }
+    // IonSpigot end
     protected void a(MovingObjectPosition movingobjectposition) {
         if (!this.world.isClientSide) {
             List list = Items.POTION.h(this.item);
-- 
2.29.2.windows.2

