From 9e8587194ec73f4a09a03bab164d9da2a7f157d4 Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Sat, 19 Dec 2020 18:44:26 +0000
Subject: [PATCH] Lag Compensated Ticking


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 87fe16c80..1c08a712a 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -136,6 +136,10 @@ public abstract class Entity implements ICommandListener {
     public boolean forceExplosionKnockback; // CraftBukkit - SPIGOT-949
     public boolean inUnloadedChunk = false; // PaperSpigot - Remove entities in unloaded chunks
     public boolean loadChunks = false; // PaperSpigot - Entities can load chunks they move through and keep them loaded
+    // IonSpigot start - Lag Compensated Ticking
+    protected boolean compensated;
+    protected void tick() {}
+    // IonSpigot end
 
     // Spigot start
     public Timing tickTimer = SpigotTimings.getEntityTimings(this); // Spigot
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index b029dbd55..91f01fba6 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -53,6 +53,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public int ping;
     public boolean viewingCredits;
     public VisualSettings visualSettings = new VisualSettings(); // IonSpigot - Visual Settings API
+    public List<Entity> lagCompensatedTicking = new ArrayList<>(); // IonSpigot - Lag Compensated Ticking
 
     // CraftBukkit start
     public String displayName;
@@ -300,6 +301,21 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public void l() {
         try {
             super.t_();
+            // IonSpigot start - Lag Compensated Ticking
+            for (int i = 0; i < this.lagCompensatedTicking.size(); ++i) {
+                Entity entity = this.lagCompensatedTicking.get(i);
+                entity.tick();
+
+                // Check if size is > 9, this should cover some abuse
+                if (entity.dead || this.lagCompensatedTicking.size() > 9) {
+                    if (!entity.dead) {
+                        entity.compensated = false;
+                    }
+
+                    this.lagCompensatedTicking.remove(i--);
+                }
+            }
+            // IonSpigot end
 
             for (int i = 0; i < this.inventory.getSize(); ++i) {
                 ItemStack itemstack = this.inventory.getItem(i);
-- 
2.29.2.windows.2

