From ac882b5bc711da186e7389901acccb3ca521e952 Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Wed, 16 Dec 2020 18:26:12 +0000
Subject: [PATCH] Lag Compensated Potions


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index fc199cfed..0f6d824fb 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -149,4 +149,9 @@ public class IonWorldConfig {
         movementEntityCollisions = getBoolean("movement.entity-collisions", true);
     }
 
+    public boolean lagCompensatedPotions;
+    private void LagCompensatedPotions() {
+        lagCompensatedPotions = getBoolean("potions.lag-compensated", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityPotion.java b/src/main/java/net/minecraft/server/EntityPotion.java
index fd174c346..567f4797c 100644
--- a/src/main/java/net/minecraft/server/EntityPotion.java
+++ b/src/main/java/net/minecraft/server/EntityPotion.java
@@ -25,6 +25,12 @@ public class EntityPotion extends EntityProjectile {
     public EntityPotion(World world, EntityLiving entityliving, ItemStack itemstack) {
         super(world, entityliving);
         this.item = itemstack;
+        // IonSpigot start - Lag Compensated Potions
+        if (entityliving instanceof EntityPlayer && world.ionConfig.lagCompensatedPotions) {
+            ((EntityPlayer) entityliving).lagCompensatedTicking.add(this);
+            compensated = true;
+        }
+        // IonSpigot end
     }
 
     public EntityPotion(World world, double d0, double d1, double d2, ItemStack itemstack) {
@@ -60,6 +66,17 @@ public class EntityPotion extends EntityProjectile {
         return this.item.getData();
     }
 
+    // IonSpigot start - Lag Compensated Potions
+    @Override
+    public void t_() {
+        if (!compensated) {
+            tick();
+        }
+    }
+    public void tick() {
+        super.t_();
+    }
+    // IonSpigot end
     protected void a(MovingObjectPosition movingobjectposition) {
         if (!this.world.isClientSide) {
             List list = Items.POTION.h(this.item);
-- 
2.29.2.windows.2

