From efce7ef9dd5bedeaed7f21ee6f265028541d089f Mon Sep 17 00:00:00 2001
From: "kfian294ma4@gmail.com" <kfian294ma4@gmail.com>
Date: Wed, 16 Dec 2020 18:38:26 +0000
Subject: [PATCH] Smooth Potting


diff --git a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
index 517480321..5c967779c 100644
--- a/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
+++ b/src/main/java/me/suicidalkids/ion/IonWorldConfig.java
@@ -138,4 +138,9 @@ public class IonWorldConfig {
         lagCompensatedPotions = getBoolean("potions.lag-compensated", true);
     }
 
+    public boolean smoothPotting;
+    private void SmoothPotting() {
+        smoothPotting = getBoolean("potions.smooth", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index e7cca3c6c..f0e24b36f 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -131,6 +131,12 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
                     AxisAlignedBB axisalignedbb = entity1.getBoundingBox().grow((double) f, (double) f, (double) f);
                     MovingObjectPosition movingobjectposition1 = axisalignedbb.a(vec3d, vec3d1);
 
+                    // IonSpigot start - Smooth Potting
+                    if (this instanceof EntityPotion && world.ionConfig.smoothPotting &&
+                        movingobjectposition1 == null && getBoundingBox().b(entity1.getBoundingBox())) {
+                        movingobjectposition1 = new MovingObjectPosition(entity1);
+                    }
+                    // IonSpigot end
                     if (movingobjectposition1 != null) {
                         double d1 = vec3d.distanceSquared(movingobjectposition1.pos);
 
-- 
2.29.2.windows.2

